### PR TITLE
deps(yargs): upgrade to 17.3.1

### LINKED ISSUE #13589: `--cli-flags-path` no longer works
```json
{
  "budgets": [{"resourceSizes": [{"resourceType": "script", "budget": 1}]}]
}
```

```sh
yarn start http://localhost:9009/app/ --quiet --cli-flags-path=settings.json
```

Settings `budgets` will be undefined.

I bisected to #13045. The PR doesn't update the `yargs` dependency, but did make some (seemingly trivial) modifications to `lighthouse-cli/cli-flags.js`.

Blocking https://github.com/GoogleChrome/lighthouse-ci/pull/726

### PR DESCRIPTION
**Pull request by connorjclark**
Fixes #13589

Not sure on the root cause, but I noticed that the bug didn't happen with the latest yargs.

https://github.com/yargs/yargs/blob/main/CHANGELOG.md

Only necessary changes were for:

> coerce: coerce is now applied before validation.

Which required having coerce fns return `null` (or `undefined`, based on what the settings type expected) to prevent some fields like `skipAudits`, `extraHeaders` from being `undefined` instead of `null` ... which was fine, but was an unnecessary change to the settings normalization.

### CODE REVIEW

File: lighthouse-cli/cli-flags.js [hunk 1]
```diff
@@ -352,10 +352,10 @@ function getFlags(manualArgv, options = {}) {
 /**
  * Support comma-separated values for some array flags by splitting on any ',' found.
  * @param {Array<string>=} strings
- * @return {Array<string>|undefined}
+ * @return {Array<string>|null}
  */
 function splitCommaSeparatedValues(strings) {
-  if (!strings) return strings;
+  if (!strings) return null;
```
Comment (adamraine) [inline lighthouse-cli/cli-flags.js:358]: Why do we return `null` in some spots and `undefined` in others?

File: lighthouse-core/runner.js [hunk 2]
```diff
@@ -259,10 +259,19 @@ class Runner {
...
+        skipAudits: undefined,
+        onlyAudits: undefined,
+        onlyCategories: undefined,
+        extraHeaders: undefined,
       };
       const normalizedGatherSettings = Object.assign({}, artifacts.settings, overrides);
       const normalizedAuditSettings = Object.assign({}, settings, overrides);
 
+      for (const k of Object.keys(normalizedGatherSettings)) {
+        if (!isDeepEqual(normalizedGatherSettings[k], normalizedAuditSettings[k])) {
```
Comment (adamraine) [inline lighthouse-core/runner.js:271]: Why doesn't `isDeepEqual` cover this in the existing code? Is it just to get a more specific error message?

Comment (connorjclark) [reply lighthouse-core/runner.js:271]: Yea, I could remove it but it helped a bit with debugging sooo

**Comment by adamraine** [comment]
Looks like we need to handle `null` here:

https://github.com/GoogleChrome/lighthouse/blob/f59499a50f4986dd2c53d8e46c11b8628e8a2451/lighthouse-core/config/config-helpers.js#L81

File: lighthouse-core/runner.js (hunk 2 above)
Comment (adamraine) [reply lighthouse-core/runner.js:271]: Could you add a comment about why we need two, like "First one checks equality of keys, second one checks for missing keys".

**Comment by connorjclark** [comment]
ok let's try to get `cli-flags` to return the exact same thing even with this version upgrade. We'll need some tests in cli-flags that assert things like keys not being present (as opposed to being null or explicitly undefined)
_____

master

```sh
node lighthouse-cli/index.js https://www.example.com
```

```js
{
  _: [ 'https://www.example.com' ],
  verbose: false,
  quiet: false,
  'save-assets': false,
  saveAssets: false,
  'list-all-audits': false,
  listAllAudits: false,
  'list-locales': false,
  listLocales: false,
  'list-trace-categories': false,
  listTraceCategories: false,
  'print-config': false,
  printConfig: false,
  'fraggle-rock': false,
  fraggleRock: false,
  'chrome-flags': '',
  chromeFlags: '',
  port: 0,
  hostname: '127.0.0.1',
  'enable-error-reporting': undefined,
  enableErrorReporting: undefined,
  output: [ 'html' ],
  view: false,
  channel: 'cli',
  'chrome-ignore-default-flags': false,
  chromeIgnoreDefaultFlags: false,
  '$0': 'lighthouse-cli/index.js'
}
```

new yargs, _no_ changes to cli-flags
```js
Invalid value: Argument 'screenEmulation' must be an object, specified per-property ('screenEmulation.width', 'screenEmulation.deviceScaleFactor', etc)

Specify --help for available options
```

new yargs, current PR state
```js
{
  _: [ 'https://www.example.com' ],
  verbose: false,
  quiet: false,
  'save-assets': false,
  saveAssets: false,
  'list-all-audits': false,
  listAllAudits: false,
  'list-locales': false,
  listLocales: false,
  'list-trace-categories': false,
  listTraceCategories: false,
  'print-config': false,
  printConfig: false,
  'fraggle-rock': false,
  fraggleRock: false,
  'chrome-flags': '',
  chromeFlags: '',
  port: 0,
  hostname: '127.0.0.1',
  'enable-error-reporting': undefined,
  enableErrorReporting: undefined,
  output: [ 'html' ],
  view: false,
  channel: 'cli',
  'chrome-ignore-default-flags': false,
  chromeIgnoreDefaultFlags: false,
  '$0': 'lighthouse-cli/index.js',
  screenEmulation: undefined,
  'screen-emulation': undefined,
  emulatedUserAgent: undefined,
  'emulated-user-agent': undefined,
  'gather-mode': undefined,
  G: undefined,
  gatherMode: undefined,
  'audit-mode': undefined,
  A: undefined,
  auditMode: undefined,
  'only-audits': null,
  onlyAudits: null,
  'only-categories': null,
  onlyCategories: null,
  'skip-audits': null,
  skipAudits: null,
  locale: undefined,
  throttling: undefined,
  'extra-headers': null,
  extraHeaders: null,
  plugins: null
}
```

**Comment by connorjclark** [comment]
> Not sure on the root cause, but I noticed that the bug didn't happen with the latest yargs.

must have been this https://github.com/yargs/yargs/pull/2063#:~:text=Bug%20Fixes-,esm%20json%20import,-(%23%E2%80%8B416)%20(90f970a)

> esm json import

which means we should also update our dep on yargs-parser

File: lighthouse-cli/test/cli/__snapshots__/cli-flags-test.js.snap [hunk 3]
```diff
@@ -413,9 +456,6 @@ Object {
   "extraHeaders": Object {
     "X-Men": "wolverine",
   },
-  "extraheaders": Object {
-    "xMen": "wolverine",
-  },
```
Comment (connorjclark) [inline lighthouse-cli/test/cli/__snapshots__/cli-flags-test.js.snap:418]: seems the older yargs was overzealous in how it interprets provided config files.

File: lighthouse-cli/cli-flags.js [hunk 4]
```diff
@@ -343,19 +343,28 @@ function getFlags(manualArgv, options = {}) {
...
 
 /**
  * Support comma-separated values for some array flags by splitting on any ',' found.
  * @param {Array<string>=} strings
- * @return {Array<string>|undefined}
+ * @return {Array<string>=}
  */
 function splitCommaSeparatedValues(strings) {
-  if (!strings) return strings;
+  if (!strings) return undefined;
```
Comment (adamraine) [inline lighthouse-cli/cli-flags.js:367]: Does this actually affect anything? If not, could we match the rest of the file with `Array<string>|undefined` and `if(!strings) return`.

File: lighthouse-cli/cli-flags.js [hunk 5]
```diff
@@ -343,19 +343,28 @@ function getFlags(manualArgv, options = {}) {
...
-  const argv = parser.argv;
+  const argv = /** @type {Awaited<typeof parser.argv>} */ (parser.argv);
   const cliFlags = /** @type {typeof argv & CamelCasify<typeof argv>} */ (argv);
 
+  // yargs will return `undefined` for options that have a `coerce` function but
+  // are not actually present in the user input. Instead of passing properties
+  // explicitly set to undefined, delete them from the flags object.
+  for (const [k, v] of Object.entries(cliFlags)) {
+    // This property is meant to possibly be explicitly undefined.
+    if (k === 'enable-error-reporting' || k === 'enableErrorReporting') continue;
```
Comment (adamraine) [inline lighthouse-cli/cli-flags.js:354]: Can you add tests for the special handling of this flag?

Comment (connorjclark) [reply lighthouse-cli/cli-flags.js:354]: That's the snapshot tests tho. I don't understand the comment in this flag's options anyway, @brendankenny may know.

Comment (adamraine) [reply lighthouse-cli/cli-flags.js:354]: Hmm seems like it isn't necessary https://github.com/GoogleChrome/lighthouse/pull/11794/files#r539003671

**Review by adamraine — APPROVED** [review]
